# ASCs

================

This repository contains the source code for running the analysis described in the paper.

Instructions:
- All files run from the root directory.
- Python scripts are found in '0_Python_scripts' (all scripts are run using Python3)
- R scripts are found in '0_R_scripts'
- All python scripts are run using Python 3.
- CSVs produced by Python scripts can be found in 'Author CSVs'

Folders:
- 0 - Scripts
- 1 - EMBL files (not uploaded to this repository due to size)
- 2 - FASTA files
- 3 - PaxDB raw data and appropriate EMBL genomes
- 4 - Outputs (includes CSVs)

Python scripts:

Python scripts are categorised 1-11 as follows. A spreadsheet with more detailed script information can be found inside the '0_Python_scripts' folder.

- 1 - Filtering and obtaining FASTA files
- 2 - Dinucleotide-controlled simulation experiment
- 3 - Obtain ASC frequency information - includes relative usage
- 4 - Mollicutes analysis
- 5 - Subset analysis e.g. TGA HEGs vs TAA LEGs
- 6 - Third stops analysis
- 7 - +4T identification and tests
- 8 - Primary stop usage
- 9 - +4T-controlled simulation experiment
- 10 - T-codon enrichment scores
- 11 - Lengths and expression association
